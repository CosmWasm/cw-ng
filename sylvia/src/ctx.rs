use cosmwasm_std::{Deps, DepsMut, Empty, Env, Event, MessageInfo, MsgResponse};

/// Represantation of `reply` context received in entry point.
#[non_exhaustive]
pub struct ReplyCtx<'a, C: cosmwasm_std::CustomQuery = Empty> {
    pub deps: DepsMut<'a, C>,
    pub env: Env,
    pub gas_used: u64,
    pub events: Vec<Event>,
    pub msg_responses: Vec<MsgResponse>,
}

/// Represantation of `migrate` context received in entry point.
#[non_exhaustive]
pub struct MigrateCtx<'a, C: cosmwasm_std::CustomQuery = Empty> {
    pub deps: DepsMut<'a, C>,
    pub env: Env,
}

/// Represantation of `execute` context received in entry point.
#[non_exhaustive]
pub struct ExecCtx<'a, C: cosmwasm_std::CustomQuery = Empty> {
    pub deps: DepsMut<'a, C>,
    pub env: Env,
    pub info: MessageInfo,
}

/// Represantation of `instantiate` context received in entry point.
#[non_exhaustive]
pub struct InstantiateCtx<'a, C: cosmwasm_std::CustomQuery = Empty> {
    pub deps: DepsMut<'a, C>,
    pub env: Env,
    pub info: MessageInfo,
}

/// Represantation of `query` context received in entry point.
#[non_exhaustive]
pub struct QueryCtx<'a, C: cosmwasm_std::CustomQuery = Empty> {
    pub deps: Deps<'a, C>,
    pub env: Env,
}

/// Represantation of `sudo` context received in entry point.
#[non_exhaustive]
pub struct SudoCtx<'a, C: cosmwasm_std::CustomQuery = Empty> {
    pub deps: DepsMut<'a, C>,
    pub env: Env,
}

impl<C: cosmwasm_std::CustomQuery> ExecCtx<'_, C> {
    pub fn branch(&'_ mut self) -> ExecCtx<'_, C> {
        ExecCtx {
            deps: self.deps.branch(),
            env: self.env.clone(),
            info: self.info.clone(),
        }
    }
}

impl<C: cosmwasm_std::CustomQuery> InstantiateCtx<'_, C> {
    pub fn branch(&'_ mut self) -> InstantiateCtx<'_, C> {
        InstantiateCtx {
            deps: self.deps.branch(),
            env: self.env.clone(),
            info: self.info.clone(),
        }
    }
}

impl<C: cosmwasm_std::CustomQuery> SudoCtx<'_, C> {
    pub fn branch(&'_ mut self) -> SudoCtx<'_, C> {
        SudoCtx {
            deps: self.deps.branch(),
            env: self.env.clone(),
        }
    }
}

impl<'a, C: cosmwasm_std::CustomQuery> From<(DepsMut<'a, C>, Env)> for MigrateCtx<'a, C> {
    fn from((deps, env): (DepsMut<'a, C>, Env)) -> Self {
        Self { deps, env }
    }
}

impl<'a, C: cosmwasm_std::CustomQuery>
    From<(DepsMut<'a, C>, Env, u64, Vec<Event>, Vec<MsgResponse>)> for ReplyCtx<'a, C>
{
    fn from(
        (deps, env, gas_used, events, msg_responses): (
            DepsMut<'a, C>,
            Env,
            u64,
            Vec<Event>,
            Vec<MsgResponse>,
        ),
    ) -> Self {
        Self {
            deps,
            env,
            gas_used,
            events,
            msg_responses,
        }
    }
}

impl<'a, C: cosmwasm_std::CustomQuery> From<(DepsMut<'a, C>, Env, MessageInfo)> for ExecCtx<'a, C> {
    fn from((deps, env, info): (DepsMut<'a, C>, Env, MessageInfo)) -> Self {
        Self { deps, env, info }
    }
}

impl<'a, C: cosmwasm_std::CustomQuery> From<(DepsMut<'a, C>, Env, MessageInfo)>
    for InstantiateCtx<'a, C>
{
    fn from((deps, env, info): (DepsMut<'a, C>, Env, MessageInfo)) -> Self {
        Self { deps, env, info }
    }
}

impl<'a, C: cosmwasm_std::CustomQuery> From<(Deps<'a, C>, Env)> for QueryCtx<'a, C> {
    fn from((deps, env): (Deps<'a, C>, Env)) -> Self {
        Self { deps, env }
    }
}

impl<'a, C: cosmwasm_std::CustomQuery> From<(DepsMut<'a, C>, Env)> for SudoCtx<'a, C> {
    fn from((deps, env): (DepsMut<'a, C>, Env)) -> Self {
        Self { deps, env }
    }
}
