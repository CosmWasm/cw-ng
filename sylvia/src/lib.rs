#![cfg_attr(docsrs, feature(doc_cfg))]
//! Framework for creating CosmWasm Smart Contract with high-level abstraction layer
//!
//! Most of implementation lies in `sylvia-derive` crate which is reexported here

pub mod builder;
pub mod into_response;
#[cfg_attr(docsrs, doc(cfg(feature = "mt")))]
#[cfg(feature = "mt")]
pub mod multitest;
pub mod types;
pub mod utils;

#[cfg_attr(docsrs, doc(cfg(feature = "mt")))]
#[cfg(feature = "mt")]
pub use anyhow;
#[cfg_attr(docsrs, doc(cfg(feature = "mt")))]
#[cfg(feature = "mt")]
pub use cw_multi_test;
#[cfg_attr(docsrs, doc(cfg(feature = "cosmwasm_1_2")))]
#[cfg(feature = "cosmwasm_1_2")]
pub use cw_utils;
#[cfg_attr(docsrs, doc(cfg(feature = "mt")))]
#[cfg(feature = "mt")]
pub use derivative;
pub use sylvia_derive::{contract, entry_points, interface};
pub use {
    cosmwasm_schema as cw_schema, cosmwasm_std as cw_std, schemars, serde,
    serde_cw_value as serde_value, serde_json_wasm as serde_json,
};

pub mod replies {
    use cosmwasm_std::{DepsMut, Empty, Env, Event, MsgResponse};

    /// Represantation of `reply` context received in entry point.
    #[deprecated(
        since = "1.3.0",
        note = "This type is added temporarily to not break existing API. Since `2.0.0` it will replace the `sylvia::types::ReplyCtx` type."
    )]
    #[non_exhaustive]
    pub struct ReplyCtx<'a, C: cosmwasm_std::CustomQuery = Empty> {
        pub deps: DepsMut<'a, C>,
        pub env: Env,
        pub gas_used: u64,
        pub events: Vec<Event>,
        pub msg_responses: Vec<MsgResponse>,
    }

    #[allow(deprecated)]
    impl<'a, C: cosmwasm_std::CustomQuery>
        From<(DepsMut<'a, C>, Env, u64, Vec<Event>, Vec<MsgResponse>)> for ReplyCtx<'a, C>
    {
        fn from(
            (deps, env, gas_used, events, msg_responses): (
                DepsMut<'a, C>,
                Env,
                u64,
                Vec<Event>,
                Vec<MsgResponse>,
            ),
        ) -> Self {
            Self {
                deps,
                env,
                gas_used,
                events,
                msg_responses,
            }
        }
    }
}
