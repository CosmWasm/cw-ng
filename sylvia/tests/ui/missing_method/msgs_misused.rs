#![allow(unused_imports)]
use sylvia::ctx::InstantiateCtx;
use sylvia::cw_std::{Response, StdResult};

pub struct Contract;

mod interface {
    use sylvia::ctx::{InstantiateCtx, MigrateCtx};
    use sylvia::cw_std::{Response, StdError, StdResult};

    #[sylvia::interface]
    trait Interface {
        type Error: From<StdError>;

        #[sv::msg(instantiate)]
        fn instantiate(&self, ctx: InstantiateCtx) -> StdResult<Response>;

        #[sv::msg(migrate)]
        fn migrate(&self, ctx: MigrateCtx) -> StdResult<Response>;
    }
}

#[sylvia::contract]
impl Contract {
    pub const fn new() -> Self {
        Contract
    }

    #[sv::msg(instantiate)]
    pub fn instantiate(&self, ctx: InstantiateCtx) -> StdResult<Response> {
        Ok(Response::new())
    }

    #[sv::msg(instantiate)]
    pub fn instantiate2(&self, ctx: InstantiateCtx) -> StdResult<Response> {
        Ok(Response::new())
    }
}

fn main() {}
