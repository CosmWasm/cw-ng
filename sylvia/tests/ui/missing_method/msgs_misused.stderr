error: The message attribute `instantiate` is not supported in interfaces.

         = note: Contracts need to implement `instantiate` method within their `impl` block.

  --> tests/ui/missing_method/msgs_misused.rs:16:12
   |
16 |         fn instantiate(&self, ctx: InstantiateCtx) -> StdResult<Response>;
   |            ^^^^^^^^^^^

error: The message attribute `migrate` is not supported in interfaces

         = note: Contracts need to implement `migrate` method within their `impl` block.

  --> tests/ui/missing_method/msgs_misused.rs:19:12
   |
19 |         fn migrate(&self, ctx: MigrateCtx) -> StdResult<Response>;
   |            ^^^^^^^

error: More than one instantiation or migration message

         = note: Instantiation/Migration message previously defined here

  --> tests/ui/missing_method/msgs_misused.rs:30:12
   |
30 |     pub fn instantiate(&self, ctx: InstantiateCtx) -> StdResult<Response> {
   |            ^^^^^^^^^^^
